draft-irtf-cfrg-opaque.md

```diff
@@ -1056,225 +1056,60 @@ Steps:
-- `ke3, server_info, session_key = ClientFinish(password, client_identity, server_identity, ke1, ke2)`
-- `session_key = ServerFinish(ke3)`
+This section describes the authenticated key exchange protocol for OPAQUE using 3DH,
+a 3-message AKE which satisfies the forward secrecy and KCI properties discussed in {{security-considerations}}.
+As shown in {{OPAQUE}}, OPAQUE cannot use less than three messages, so the 3-message
+instantiations presented here are optimal in terms of number of messages. On the other
+hand, there is no impediment to using OPAQUE with protocols with more than 3 messages
+as in the case of IKEv2 (or the underlying SIGMA-R protocol {{SIGMA}}).
 
-{{opaque-client}} and {{opaque-server}} specify the inner working of these functions.
+The generic outline of OPAQUE with a 3-message AKE protocol includes three messages
+KE1, KE2, and KE3, where KE1 and KE2 include key exchange shares, e.g., DH values, sent
+by the client and server, respectively, and KE3 provides explicit client authentication and
+full forward security (without it, forward secrecy is only achieved against eavesdroppers
+which is insufficient for OPAQUE security). These messages are computed and exchanged with
+the following application APIs:
 
-The `ke1`, `ke2`, and `ke3` output values are the protocol messages sent between client and server.
+- ke1, client_state = ClientInit(client_identity, password, client_info)
+- ke2, client_info = ServerInit(server_identity, server_private_key, server_public_key, record, credential_identifier, oprf_seed, ke1)
+- ke3, server_info, session_key, export_key = ClientFinish(password, client_identity, server_identity, ke2)
+- session_key, export_key = ServerFinish(ke3)
 
-The return value of `ServerFinalize()` indicates the client's authenticity to the server and therefore the validity of
-`client_info` produced by `Response()`, and `session_key` if valid (nil otherwise). If it returns nil, then the client
-MUST NOT be trusted and `client_info` and the internal value `session_key` MUST NOT be used.
+Outputs `ke1`, `ke2`, and `ke3` are the three protocol messages sent between client and server.
+Outputs `client_info`, `server_info` correspond to the optional information exchanged between
+client and server during the key exchange protocol. And finally, `session_key` and `export_key`
+are outputs to be consumed by applications. Applications can use `session_key` to derive
+additional keying material as needed.
 
-The following structure can be used by the client to carry internal values in a state between calls to `ClientInit()` and `ClientFinish()`:
+Both ClientFinish and ServerFinish return an error if authentication failed. In this case,
+clients and servers MUST NOT use any outputs from the protocol, such as `session_key` or
+`export_key`. ClientInit and ServerInit both implicitly return internal state objects
+`client_state` and `server_state`, respectively, with the following named fields:
```

Comment:
If it's implicit, then maybe not return it at line 1075 :) (or add it at line 1076)